lib/zmq-kernel.js

```diff
@@ -194,13 +203,17 @@ export default class ZMQKernel extends Kernel {
     );
     this.kernelProcess = spawn;
     this.monitor(() => {
-      if (onRestarted) onRestarted(this);
+      if (onRestarted) onRestarted();
     });
   }
 
   // onResults is a callback that may be called multiple times
   // as results come in from the kernel
-  _execute(code: string, requestId: string, onResults: ?Function) {
+  execute(code: string, callWatches: boolean, onResults: ResultsCallback) {
+    log("ZMQKernel.execute:", code, callWatches);
+    // requestId starting with "execute_" will trigger watches being called
+    const requestId = callWatches ? `execute_${v4()}` : `watch_${v4()}`;
```

Comment:
neat!

Reply:
though, since `callWatches` is used less often, I'd change the signature to `execute(code: string, onResults: ResultsCallback, callWatches: boolean)` so that there is no need to update all the calls to `execute` throughout Hydrogen.

Reply:
In terms of cascading effects, there aren't any because `kernel.js` still defines both `execute` and `executeWatch` (for now).

I'm not sure that switching argument order in the public plugin API would have a positive effect. Here's some code I have from a plugin I wrote: 
```
next.execute(code, callWatches, (msg, channel) => {
  if (!makeReply && msg.parent_header) {
    makeReply = createMessageFactory(msg.parent_header);
  }

  onResults(msg, channel);
  if (callWatches && msg.header.msg_type == "execute_reply") {
    if (this.enableVariableExplorer){
      this.variableExplorerHook(next);
    }
  }
});
```

The results callback is quite long, and note that `callWatches` is a mandatory argument because a plugin will generally passthrough that argument. That said there are definitely more details to think through wrt `callWatches`. What do you think?

Reply:
We shouldn't make the watches callback public. Especially, since ultimately, we want to implement watches using a plugin.

Reply:
Of course I'd be happy to hear any ideas of how to handle this more cleanly!

Ultimately more issues in this vein might arise: for example Jupyter execute messages have a `store_history` option that may need to be toggled by plugins to exclude non-user-initiated requests.

Reply:
> So this is tricky: let's say a plugin splits one execute request into three. Then you'd only want to run watches after the last request, because the three requests are part of a single user-initiated action.

As I said above, it has an easy solution. We should think a watches as just one more plugin in the chain.

> This is a mess and should be reconsidered in the future, but I don't want to block this PR on the lack of a good solution.

No, to my understanding, this is a blocking issue. We don't want to make public an API that we don't want to support in the future (and one that would make harder the implementation of watches as a plugin).

Reply:
> We should think a watches as just one more plugin in the chain.

That only works if watches is the very first plugin in the chain. It doesn't work if it's not the first plugin.

Eventually we can introduce the notion of plugin priority (basically in addition to `execute`/`inspect`/... fields, have a `priority` field that's a number). But this is postponed for the future.

Reply:
To be clear, the plan is to eventually replace `callWatches` with an object that specifies some of the following:
* Which user-initiated action is this call part of?
* Is it the first call for this user-initiated action?
* Is it the last call for this user-initiated action?
* Are the contents of this request something from the user's code, or something generated by the plugin?
* Should history be logged with the kernel for this request?

I certainly don't want to think through all of the above in this PR.

Reply:
> That only works if watches is the very first plugin in the chain. It doesn't work if it's not the first plugin.

If watches were first, then they wouldn't be affected by any of the other plugins.

To be more precise: the request code of a watch wouldn't be affected by any of the plugins, but the response message would be affected by all the plugins.

> To be clear, the plan is to eventually replace callWatches with an object that specifies some of the following:
> [...]
> I certainly don't want to think through all of the above in this PR.

As I said above, plugins don't need to know about `callWatches`. The decision of what Hydrogen has to do with a response message should be implemented in the first onResponse callback that Hydrogen passes to the plugin chain.

Reply:
> To be more precise: the request code of a watch wouldn't be affected by any of the plugins, but the response message would be affected by all the plugins.

Not sure I understand what you said. Could you rephrase that?

To repeat my point here: let's say the user made an execute request for `foo()`. Then a plugin split that into two execute requests: `foo()` and `bar()`. If the watches plugin comes later, it would execute watches **twice**: once after `foo()` and again after `bar()`. This is wrong. Ergo watches plugin must come first, or `callWatches` boolean must exist.

Reply:
> Not sure I understand what you said. Could you rephrase that?

This was a misunderstanding on my side. We do want the chain of `onResult` callbacks affect the results received by Watches.

---

> To repeat my point here: let's say the user made an execute request for foo(). Then a plugin split that into two execute requests: foo() and bar(). If the watches plugin comes later, it would execute watches twice: once after foo() and again after bar()

I thought I had replied to this comment before, but I can't find my response. So I repeat it here: a plugin is responsible for invoking the parent's onResult callback only once.